core/executor/src/heap.rs

```diff
@@ -0,0 +1,356 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! This module implements a buddy allocation heap.
+//! It uses a binary tree and follows the concepts outlined in
+//! https://en.wikipedia.org/wiki/Buddy_memory_allocation.
+
+extern crate fnv;
+
+use std::vec;
+use self::fnv::FnvHashMap;
+
+// The pointers need to be aligned. By choosing a block size
+// which is a multiple of the memory alignment requirement
+// it is ensured that a pointer is always aligned. This is
+// because in buddy allocation a pointer always points to the
+// start of a block.
+//
+// In our case the alignment for wasm32-unknown-unknown is
+// 1 byte though, i.e. the pointer will always be aligned.
+const BLOCK_SIZE: u32 = 8192; // 2^13 Bytes
+
+#[derive(PartialEq, Copy, Clone)]
+enum Node {
+	Free,
+	Full,
+	Split,
+}
+
+/// A buddy allocation heap, which tracks allocations and deallocations
+/// using a binary tree.
+pub struct Heap {
+	allocated_bytes: FnvHashMap<u32, u32>,
+	levels: u32,
+	ptr_offset: u32,
+	tree: vec::Vec<Node>,
+	total_size: u32,
+}
+
+impl Heap {
+
+	/// Creates a new buddy allocation heap.
+	///
+	/// # Arguments
+	///
+	/// * `ptr_offset` - The pointers returned by `allocate()`
+	///   will start from this offset on.
+	/// * `heap_size` - The size available to this heap instance
+	///   (in Bytes) for allocating memory.
+	///
+	pub fn new(ptr_offset: u32, heap_size: u32) -> Self {
+		let leaves = heap_size / BLOCK_SIZE;
+		let levels = Heap::get_tree_levels(leaves);
+		let node_count: usize = (1 << levels + 1) - 1;
+
+		Heap {
+			allocated_bytes: FnvHashMap::default(),
+			levels,
+			ptr_offset,
+			tree: vec![Node::Free; node_count],
+			total_size: 0,
+		}
+	}
+
+	/// Gets requested number of bytes to allocate and returns a pointer.
+	pub fn allocate(&mut self, size: u32) -> u32 {
```

Comment:
This one doesn't seem to ensure that allocation is aligned.

`ptr_offset` is seeded by `used_size` returned by wasmi, but theoretically it is not guaranteed to be aligned properly. I think because maximum value type handled by wasm32 is `i64`, we need to make sure that pointers from this function are always aligned by 8 bytes.

Reply:
Ok, this should be ensured as long as block size is a multiple of 8 (currently it's 8192). So I think the PR is already safe for that. I could add an assertion when creating the heap and make the documentation more explicit?